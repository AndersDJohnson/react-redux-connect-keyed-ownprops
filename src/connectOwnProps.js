import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import shallowequal from 'shallowequal'

const pickOwnProps = (propNames, ownProps) =>
  typeof propNames === 'function'
    ? propNames(ownProps)
    : (Array.isArray(propNames) ? propNames : [propNames]).reduce(
        (acc, propName) => ({
          ...acc,
          [propName]: ownProps[propName]
        }),
        {}
      )

const makeMakeMapFunction = (propNames, makeOrMapFunction) => {
  if (typeof makeOrMapFunction !== 'function') return makeOrMapFunction

  const makeMapFunction = (_state, ownProps) => {
    const pickedOwnProps = pickOwnProps(propNames, ownProps)

    let mapFunction = makeOrMapFunction

    return state => {
      let result = mapFunction(state, pickedOwnProps)
      // Handle factory functions.
      if (typeof result === 'function') {
        mapFunction = result
        result = mapFunction(state, pickOwnProps)
      }
      return result
    }
  }

  return makeMapFunction
}

const makeKeyedOwnProps = (propNames, ConnectedComp) => {
  let key = 0
  let lastPropsKey
  let lastPickedOwnProps

  const ConnectOwnProps = props => {
    const pickedOwnProps = pickOwnProps(propNames, props)

    if (
      props.key !== lastPropsKey ||
      !shallowequal(pickedOwnProps, lastPickedOwnProps)
    ) {
      key++

      lastPropsKey = props.key
      lastPickedOwnProps = pickedOwnProps
    }

    return (
      <ConnectedComp {...props} key={key}>
        {props.children}
      </ConnectedComp>
    )
  }

  ConnectOwnProps.displayName = `ConnectOwnProps(${ConnectedComp.displayName})`

  ConnectOwnProps.propTypes = {
    children: PropTypes.node,
    key: PropTypes.any
  }

  return ConnectOwnProps
}

const makeConnectOwnProps = (
  propNames,
  makeMapStateToProps,
  makeMapDispatchToProps,
  mergeProps,
  options
) => Comp => {
  const ConnectedComp = connect(
    makeMapStateToProps,
    makeMapDispatchToProps,
    mergeProps,
    options
  )(Comp)

  const ConnectOwnProps = makeKeyedOwnProps(propNames, ConnectedComp)

  return ConnectOwnProps
}

const connectOwnProps = (
  propNames,
  mapStateToProps,
  mapDispatchToProps,
  mergeProps,
  options
) => {
  const makeMapStateToProps = makeMakeMapFunction(propNames, mapStateToProps)
  const makeMapDispatchToProps = makeMakeMapFunction(
    propNames,
    mapDispatchToProps
  )

  const madeConnectOwnProps = makeConnectOwnProps(
    propNames,
    makeMapStateToProps,
    makeMapDispatchToProps,
    mergeProps,
    options
  )

  return madeConnectOwnProps
}

export {
  pickOwnProps,
  makeMakeMapFunction,
  makeKeyedOwnProps,
  makeConnectOwnProps
}

export default connectOwnProps
