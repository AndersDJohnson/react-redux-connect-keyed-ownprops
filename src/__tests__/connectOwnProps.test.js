import React from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import connectOwnProps, {
  pickOwnProps,
  makeMakeMapFunction,
  makeKeyedOwnProps,
  makeConnectOwnProps
} from '../connectOwnProps'

describe('pickOwnProps', () => {
  test('picks props', () => {
    const pickedOwnProps = pickOwnProps(['myProp'], { myProp: 1 })
    expect(pickedOwnProps).toEqual({
      myProp: 1
    })
  })

  test('ignores nonexistent props', () => {
    const pickedOwnProps = pickOwnProps(['myNonexistentProp'], { myProp: 1 })
    expect(pickedOwnProps).toEqual({})
  })

  test('works with single prop', () => {
    const pickedOwnProps = pickOwnProps('myProp', { myProp: 1 })
    expect(pickedOwnProps).toEqual({
      myProp: 1
    })
  })

  test('works with function', () => {
    const pickedOwnProps = pickOwnProps(({ myProp }) => ({ myProp }), {
      myProp: 1
    })
    expect(pickedOwnProps).toEqual({
      myProp: 1
    })
  })
})

describe('makeMakeMapFunction', () => {
  test('works', () => {
    const mapFunction = state => ({
      ...state,
      mappedProp: 1
    })

    const madeMakeMapFunction = makeMakeMapFunction(['myProp'], mapFunction)

    const wrappedMapFunction = madeMakeMapFunction(
      {},
      {
        myProp: 2,
        unusedProp: 3
      }
    )

    expect(
      wrappedMapFunction({
        existingProp: 4
      })
    ).toEqual({
      mappedProp: 1,
      existingProp: 4
    })
  })

  test('works with factory', () => {
    const makeMapFunction = jest.fn((state, pickedOwnProps) => {
      const selectOwnProp = () => 0

      return state => ({
        ...state,
        selectedOwnProp: selectOwnProp(state),
        mappedProp: 1
      })
    })

    const madeMakeMapFunction = makeMakeMapFunction(['myProp'], makeMapFunction)

    const wrappedMapFunction = madeMakeMapFunction(
      {
        myState: 1
      },
      {
        myProp: 2,
        unusedProp: 3
      }
    )

    expect(
      wrappedMapFunction({
        existingProp: 4
      })
    ).toEqual({
      mappedProp: 1,
      selectedOwnProp: 0,
      existingProp: 4
    })

    expect(
      wrappedMapFunction({
        existingProp: 5
      })
    ).toEqual({
      mappedProp: 1,
      selectedOwnProp: 0,
      existingProp: 5
    })

    expect(makeMapFunction).toHaveBeenCalledTimes(1)
  })
})

const Comp = ({ myProp }) => {
  return <div>{myProp}</div>
}

Comp.propTypes = {
  myProp: PropTypes.string
}

describe('makeKeyedOwnProps', () => {
  test('works', () => {
    const MadeKeyedtOwnProps = makeKeyedOwnProps(['myProp'], Comp)

    const wrapper = mount(<MadeKeyedtOwnProps myProp="myValue" />)

    const $Comp = () => wrapper.find(Comp)

    expect($Comp().key()).toEqual('1')

    expect(wrapper).toMatchSnapshot()

    wrapper.setProps({ myProp: 'myValue' })

    expect($Comp().key()).toEqual('1')

    expect(wrapper).toMatchSnapshot()

    wrapper.setProps({ myProp: 'myValue2' })

    expect($Comp().key()).toEqual('2')

    expect(wrapper).toMatchSnapshot()
  })
})

describe('makeConnectOwnProps', () => {
  it('works', () => {
    const madeConnectOwnProps = makeConnectOwnProps(['myProp'], () => ({}))

    const ConnectOwnProps = madeConnectOwnProps(Comp)

    const store = createStore(() => {})

    const wrapper = mount(
      <Provider store={store}>
        <ConnectOwnProps myProp="myValue" />
      </Provider>
    )

    const $Comp = () => wrapper.find('Connect(Comp)')

    expect($Comp().key()).toEqual('1')

    expect(wrapper).toMatchSnapshot()

    wrapper.setProps({ children: <ConnectOwnProps myProp="myValue" /> })

    expect($Comp().key()).toEqual('1')

    expect(wrapper).toMatchSnapshot()

    wrapper.setProps({ children: <ConnectOwnProps myProp="myValue2" /> })

    expect($Comp().key()).toEqual('2')

    expect(wrapper).toMatchSnapshot()
  })
})

describe('connectOwnProps', () => {
  it('works', () => {
    const MyComponent = connectOwnProps(['myProp'], () => ({}))(Comp)

    const store = createStore(() => {})

    const wrapper = mount(
      <Provider store={store}>
        <MyComponent myProp="myValue" />
      </Provider>
    )

    const $Comp = () => wrapper.find('Connect(Comp)')

    expect($Comp().key()).toEqual('1')

    expect(wrapper).toMatchSnapshot()

    wrapper.setProps({ children: <MyComponent myProp="myValue" /> })

    expect($Comp().key()).toEqual('1')

    expect(wrapper).toMatchSnapshot()

    wrapper.setProps({ children: <MyComponent myProp="myValue2" /> })

    expect($Comp().key()).toEqual('2')

    expect(wrapper).toMatchSnapshot()
  })
})
