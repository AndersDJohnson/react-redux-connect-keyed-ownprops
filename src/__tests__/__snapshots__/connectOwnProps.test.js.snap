// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectOwnProps works 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectOwnProps(Connect(Comp))
    myProp="myValue"
  >
    <Connect(Comp)
      key="1"
      myProp="myValue"
    >
      <Comp
        dispatch={[Function]}
        myProp="myValue"
      >
        <div>
          myValue
        </div>
      </Comp>
    </Connect(Comp)>
  </ConnectOwnProps(Connect(Comp))>
</Provider>
`;

exports[`connectOwnProps works 2`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectOwnProps(Connect(Comp))
    myProp="myValue"
  >
    <Connect(Comp)
      key="1"
      myProp="myValue"
    >
      <Comp
        dispatch={[Function]}
        myProp="myValue"
      >
        <div>
          myValue
        </div>
      </Comp>
    </Connect(Comp)>
  </ConnectOwnProps(Connect(Comp))>
</Provider>
`;

exports[`connectOwnProps works 3`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectOwnProps(Connect(Comp))
    myProp="myValue2"
  >
    <Connect(Comp)
      key="2"
      myProp="myValue2"
    >
      <Comp
        dispatch={[Function]}
        myProp="myValue2"
      >
        <div>
          myValue2
        </div>
      </Comp>
    </Connect(Comp)>
  </ConnectOwnProps(Connect(Comp))>
</Provider>
`;

exports[`makeConnectOwnProps works 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectOwnProps(Connect(Comp))
    myProp="myValue"
  >
    <Connect(Comp)
      key="1"
      myProp="myValue"
    >
      <Comp
        dispatch={[Function]}
        myProp="myValue"
      >
        <div>
          myValue
        </div>
      </Comp>
    </Connect(Comp)>
  </ConnectOwnProps(Connect(Comp))>
</Provider>
`;

exports[`makeConnectOwnProps works 2`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectOwnProps(Connect(Comp))
    myProp="myValue"
  >
    <Connect(Comp)
      key="1"
      myProp="myValue"
    >
      <Comp
        dispatch={[Function]}
        myProp="myValue"
      >
        <div>
          myValue
        </div>
      </Comp>
    </Connect(Comp)>
  </ConnectOwnProps(Connect(Comp))>
</Provider>
`;

exports[`makeConnectOwnProps works 3`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectOwnProps(Connect(Comp))
    myProp="myValue2"
  >
    <Connect(Comp)
      key="2"
      myProp="myValue2"
    >
      <Comp
        dispatch={[Function]}
        myProp="myValue2"
      >
        <div>
          myValue2
        </div>
      </Comp>
    </Connect(Comp)>
  </ConnectOwnProps(Connect(Comp))>
</Provider>
`;
